macrosAutoCats/src/main/scala/com/softwaremill/macwire/autocats/MacwireAutoCatsMacros.scala

```diff
@@ -17,144 +19,292 @@ object MacwireCatsEffectMacros {
     val targetType = implicitly[c.WeakTypeTag[T]]
     lazy val typeCheckUtil = new TypeCheckUtil[c.type](c, log)
 
-    trait Provider {
+    val providersEntities = new Providers[c.type](c)
+    import providersEntities._
+
+    def providerFromExpr(expr: Expr[Any]): Provider = {
+      val tree = expr.tree
+      (Resource.fromTree(tree) orElse Effect.fromTree(tree) orElse FactoryMethod.fromTree(tree))
+        .getOrElse(new Instance(tree))
+    }
+
+    def findProviderIn(values: Seq[Value])(tpe: Type): Option[Tree] = values.find(_.`type` <:< tpe).map(_.ident)
+
+    val providers = dependencies.map(providerFromExpr)
+    
+    val values: Seq[Value] = {
+
+      val (fms, initValues): (Vector[FactoryMethod], Vector[Value]) = providers.toVector.partitionBifold {
+        case e: Effect => Right(e.asResource)
+        case v: Value  => Right(v)
+        case fm: FactoryMethod        => Left(fm)
+      }
+
+
+      def freshInstanceFromEmptyConstructor(t: Type): Option[Tree] = {
+        (ConstructorCrimper.constructorTreeV2(c, log)(t, (_, _) => None) orElse CompanionCrimper
+        .applyTreeV2(c, log)(t, (_, _) => None))
+      }
+
+      def go(fms: Vector[FactoryMethod], values: Vector[Value]): Vector[Value] = fms match {
+        case Vector() => values
+        case v => {
+          log(s"Resolving for FMs [${fms.mkString(", ")}] with values [${values.mkString(", ")}]")
+          //FIXME we need to make use of empty constructors at this point :/
+          val (remainingFms, appliedFms) = v.partitionBifold(f => f.maybeResult(t => findProviderIn(values)(t).orElse(freshInstanceFromEmptyConstructor(t))).toRight(f))
```

Comment:
This is a very brute-force sorting algorithm ;) Aside from performance (n^2) and the (fixable?) need for no-arg constructors, there's another flaw that's directly impacting user experience - we can't let the users know which dependency is missing, and what's the path.

That's why I'd still insist on creating a proper dependency graph (vertices - providers, edges - dependencies) and sorting the vertices topologically, which would give us the proper creation order, plus a nice way to report errors.